src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,467 @@
+            SUPPRESSED_FIELDS_IN_UTIL_CLASSES, "be suppressed");
+
+        final ArchRule fieldsShouldBeImmutable = fields()
+            .that()
+            .areDeclaredInClassesThat()
+            .haveSimpleNameEndingWith("Util")
+            .should(BE_IMMUTABLE)
+            .andShould()
+            .beFinal()
+            .andShould()
+            .beStatic()
+            .orShould(beSuppressed);
+
+        fieldsShouldBeImmutable.check(utilClasses);
+    }
+
+    /**
+     * Test to ensure modules annotated with {@link StatelessCheck} contain immutable fields
+     * and modules with mutable fields are annotated with {@link FileStatefulCheck} or
+     * {@link GlobalStatefulCheck}.
+     *
+     * @noinspection JUnitTestMethodWithNoAssertions
+     * @noinspectionreason JUnitTestMethodWithNoAssertions - asserts in callstack,
+     *     but not in this method
+     */
+    @Test
+    public void testModuleImmutability() {
+        final JavaClasses checkstyleChecks = new ClassFileImporter()
+            .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS)
+            .importPackages("com.puppycrawl.tools.checkstyle.checks");
+
+        final ModulePropertyPredicate moduleProperties = new ModulePropertyPredicate();
+
+        final DescribedPredicate<JavaClass> immutableFields = new ImmutableFieldsPredicate();
+
+        final ArchCondition<JavaClass> beAddedToSuppressionList =
+            new JavaClassSuppressionArchCondition(
+                SUPPRESSED_CLASSES, "be added to the suppression list");
+
+        final ArchCondition<JavaField> beSuppressed = new JavaFieldSuppressionArchCondition(
+            SUPPRESSED_FIELDS_IN_MODULES, "be suppressed");
+
+        final ArchCondition<JavaField> bePrivateExceptMessageKeys =
+            new ArchCondition<>("be private") {
+
+                @Override
+                public void check(JavaField item, ConditionEvents events) {
+                    if (!item.getName().startsWith("MSG")
+                        && !item.getModifiers().contains(JavaModifier.PRIVATE)) {
+                        final String message = String
+                            .format(Locale.ROOT, "Field <%s> should %s in %s",
+                                    item.getFullName(), getDescription(),
+                                    item.getSourceCodeLocation().toString());
+                        events.add(SimpleConditionEvent.violated(item, message));
+                    }
+                }
+            };
+
+        final ArchRule fieldsInStatelessChecksAreImmutable = fields()
+            .that()
+            .areDeclaredInClassesThat()
+            .areAnnotatedWith(StatelessCheck.class)
+            .and(are(not(moduleProperties)))
+            .should(BE_IMMUTABLE)
+            .andShould()
+            .beFinal()
+            .andShould(bePrivateExceptMessageKeys)
+            .orShould(beSuppressed);
```

Comment:
Let's try to keep conditions as simple as possible:

```suggestion
            .andShould(bePrivate)
            .orShould(beMessageKeys)
            .orShould(beSuppressed);
```

Reply:
Both conditions are combined as `orShould(..)` will result in `(a && b && c) || d`, whereas we would like `a && b && (c || d)`.
```java
final ArchRule fieldsInStatelessChecksShouldBeImmutable = fields()                 
    .that()                                                                        
    .areDeclaredInClassesThat()                                                    
    .areAnnotatedWith(StatelessCheck.class)                                        
    .and(are(not(moduleProperties)))                                               
    .should(BE_IMMUTABLE)                                                          
    .andShould()                                                                   
    .beFinal()                                                                     
    .andShould()                                                                   
    .bePrivate()                                                                   
    .orShould()                                                                    
    .haveNameStartingWith("MSG")                                                   
    .orShould(beSuppressedField);                                                 
```
will not give a violation for 
```
public static String MSG_KEY = "variable.declaration.usage.distance";
```
whereas the current `ArchRule` will.